File: Kruskal_Algorithm/Kruskal_Algorithm.cs
/*Kruskal's algorithm is a minimum-spanning-tree algorithm which finds an 
edge of the least possible weight that connects any two trees in the forest

Code in C#
*/

using System;

public struct Edge
{
    public int Source;
    public int Destination;
    public int Weight;
}

public struct Graph
{
    public int VerticesCount;
    public int EdgesCount;
    public Edge[] edge;
}

public struct Subset
{
    public int Parent;
    public int Rank;
}

public static Graph CreateGraph(int verticesCount, int edgesCoun)
{
    Graph graph = new Graph();
    graph.VerticesCount = verticesCount;
    graph.EdgesCount = edgesCoun;
    graph.edge = new Edge[graph.EdgesCount];

    return graph;
}

private static int Find(Subset[] subsets, int i)
{
    if (subsets[i].Parent != i)
	    subsets[i].Parent = Find(subsets, subsets[i].Parent);

    return subsets[i].Parent;
}

private static void Union(Subset[] subsets, int x, int y)
{
    int xroot = Find(subsets, x);
    int yroot = Find(subsets, y);

    if (subsets[xroot].Rank < subsets[yroot].Rank)
	    subsets[xroot].Parent = yroot;
    else if (subsets[xroot].Rank > subsets[yroot].Rank)
	    subsets[yroot].Parent = xroot;
    else
    {
         subsets[yroot].Parent = xroot;
         ++subsets[xroot].Rank;
    }
}

private static void Print(Edge[] result, int e)
{
    for (int i = 0; i < e; ++i)
        Console.WriteLine("{0} -- {1} == {2}", result[i].Source, result[i].Destination, result[i].Weight);
}

public static void Kruskal(Graph graph)
{
    int verticesCount = graph.VerticesCount;
    Edge[] result = new Edge[verticesCount];
    int i = 0;
    int e = 0;

    Array.Sort(graph.edge, delegate (Edge a, Edge b)
        return a.Weight.CompareTo(b.Weight);
    );

    Subset[] subsets = new Subset[verticesCount];

    for (int v = 0; v < verticesCount; ++v)
    {
	    subsets[v].Parent = v;
	    subsets[v].Rank = 0;
    }

    while (e < verticesCount - 1)
    {
	    Edge nextEdge = graph.edge[i++];
	    int x = Find(subsets, nextEdge.Source);
	    int y = Find(subsets, nextEdge.Destination);

	    if (x != y)
	    {
		    result[e++] = nextEdge;
		    Union(subsets, x, y);
	    }
    }

    Print(result, e);
}

/*
Input
=int verticesCount = 4;
int edgesCount = 5;
Graph graph = CreateGraph(verticesCount, edgesCount);

// Edge 0-1
graph.edge[0].Source = 0;
graph.edge[0].Destination = 1;
graph.edge[0].Weight = 10;

// Edge 0-2
graph.edge[1].Source = 0;
graph.edge[1].Destination = 2;
graph.edge[1].Weight = 6;

// Edge 0-3
graph.edge[2].Source = 0;
graph.edge[2].Destination = 3;
graph.edge[2].Weight = 5;

// Edge 1-3
graph.edge[3].Source = 1;
graph.edge[3].Destination = 3;
graph.edge[3].Weight = 15;

// Edge 2-3
graph.edge[4].Source = 2;
graph.edge[4].Destination = 3;
graph.edge[4].Weight = 4;

Kruskal(graph);

Output
2 -- 3 == 4
0 -- 3 == 5
0 -- 1 == 10

*/

File: Kruskal_Algorithm/Kruskal_Algorithm.dart
/*
Kruskal's algorithm is a minimum spanning tree algorithm that takes a graph a
s input and finds the subset of the edges of that graph which

(1) form a tree that includes every vertex
(2) has the minimum sum of weights among all the trees that can be
    formed from the graph
 */

import 'dart:io';

var INT_MAX = 9223372036854775807;

int find(i, parent)
{
    while (parent[i] != i) {
        i = parent[i];
    }
    return i;
}

void union1(i, j, parent)
{
    var a = find(i, parent);
    var b = find(j, parent);

    parent[a] = b;
}

void kruskal_mst(adjmat, V)
{
    var mincost = 0;

    var parent = new List(V);

    for (var i = 0; i < V; i++)
    {
        parent[i] = i;
    }

    var edge_count = 0;
    while (edge_count < V - 1)
    {
        var min = INT_MAX, a = -1, b = -1;

        for (var i = 0; i < V; i++)
        {
            for (var j = 0; j < V; j++)
            {
                if (find(i, parent) != find(j, parent) && adjmat[i][j] < min)
                {
                    min = adjmat[i][j];
                    a = i;
                    b = j;
                }
            }
        }

        union1(a, b, parent);
        print('Edge ${edge_count++}:(${a}, ${b}) cost:${min} \n');

        mincost += min;
    }

    print('\n Minimum cost= ${mincost} \n');
}

int main()
{
    print('Enter number of nodes 0 to ?');

    var n = int.parse(stdin.readLineSync());
    var max_edges = (n + 1) * (n);
    var adjmat = new List.generate(n+1, (_) => new List(n + 1));

    for(var i = 0; i <= n; i++)
    {
        for(var j = 0; j <= n; j++)
        {
            adjmat[i][j] = INT_MAX;
        }
    }

    print('Enter in the following format\nsrc\ndest\nweight\n');
    for(var i = 0; i < max_edges; i++)
    {
        var src = int.parse(stdin.readLineSync());
        var dest = int.parse(stdin.readLineSync());
        var weight = int.parse(stdin.readLineSync());

        print('*' * 20);

        if((src == -1) && (dest == -1))
        {
            break;
        }

        if(src > n || dest > n || src < 0 || dest < 0)
        {
            print('Invalid edge!\n');
            i--;
        }

        else
        {
            adjmat[src][dest] = weight;
        }
    }

    kruskal_mst(adjmat , n + 1);
    return 0;
}

/*
Input:
Enter number of nodes 0 to ?
4
Enter in the following format
Source
Destination
Weight

 Let the given graph is :

     (1)____1___(2)
    /  \       /  \
   3    4     4    6
  /      \   /      \
 /        \ /        \
(0)___5___(5)____5___(3)
 \         |         /
  \        |        /
   \       |       /
    \      2      /
     6     |     8
      \    |    /
       \   |   /
        \  |  /
         \ | /
          (4)

     adjmat = [
        [ 0, 3, 0, 0, 6, 5 ],
        [ 3, 0, 1, 0, 0, 4 ],
        [ 0, 1, 0, 6, 0, 4 ],
        [ 0, 0, 6, 0, 8, 5 ],
        [ 6, 0, 0, 8, 0, 2 ],
        [ 5, 4, 4, 5, 2, 0 ]
    ];
Output:

 Edge: 0-1 cost: 3
 Edge: 1-2 cost: 1
 Edge: 1-5 cost: 4
 Edge: 5-4 cost: 2
 Edge: 5-3 cost: 5
 Minimum Weight is 15
 */

